### PR TITLE
feat: Production Creating

### LINKED ISSUE #240: Submit draft production
Add gql mutation so production manager can submit their production for approval.

1. Check requirements listed below are met if issue return field errors (where applicable) for the issues
2. Set the production status to `SUBMITTED` if requirements met

Requirements:
- Check non-null fields - all fields are required except `Facebook url`, `age rating` and `cover image`
- Check it has at least one performance

### PR DESCRIPTION
**Pull request by atoff**
TODO:

- [x] Get image uploading in
- [x] Check user can set society on production mutation
- [x] Mutations to allow editing of production user's permissions
- [x] Discount mutations
- [x] Production status change mutation + ability
- [x] Add status after pending called approved. Comp bookings should only be able to be created from then, and listing should be hidden / not bookable. Production team should be able to move from approved into published **THIS NEEDS TO BE CHANGED ON FILTERS**
- [x] Allow for message pass through on status change (e.g. a review comment)
- [x] We might want to show productions via queries that have been approved? (What if it is approved, give comp booking, that user doesn't have perms, can't see production details!)
- [x] What happens on SetProductionStatus if the production id passed doesn't resolve to production
- [x] Test email sending
- [x] Send emails to approvers when submitted?
- [x] Validate the facebookEvent url on production model
- [x] Add protection to the image upload endpoint
- [x] Query for societies that user can create productions for?
- [x] Check permissions actually come through
- [x] Once booking-query-vun merged, add new view_bookings permission into create and permissionsmeta
- [x] Allow people to update their production's slug
- [x] Only allow transition to closed once production finished?

Closes #240 #241 #242 #239

### CODE REVIEW

File: uobtheatre/discounts/schema.py [hunk 1]
```diff
@@ -30,3 +39,21 @@ def get_queryset(cls, queryset, info):
...
+        update_ability = ModifyConcessionType
+
+
+class DeleteConcessionTypeMutation(ModelDeletionMutation):
+    class Meta:
+        model = ConcessionType
+        ability = ModifyConcessionType
+
+
+class Mutation(AuthMutation, graphene.ObjectType):
```
Comment (JElgar) [inline uobtheatre/discounts/schema.py:57]:
```suggestion
class Mutation(graphene.ObjectType):
```

File: uobtheatre/productions/models.py [hunk 2]
```diff
@@ -279,6 +283,8 @@ class Meta:
         permissions = (
             ("boxoffice", "Can use boxoffice for this production"),
             ("sales", "Can view sales for this production"),
+            ("force_change_production", "Can edit production once live"),
+            ("approve_production", "Can edit production once live"),
```
Comment (JElgar) [inline uobtheatre/productions/models.py:287]: Wrong description

File: uobtheatre/productions/schema.py [hunk 3]
```diff
@@ -343,6 +376,79 @@ class Meta:
...
+        intended_production = cls.get_python_value(root, info, "production", **mInput)
+        if not cls.is_creation:
+            # Check the user can access it currently
+            production = cls.get_object_instance(root, info, **mInput).production
+        else:
+            # Check user can access intended production
+            if not intended_production:
+                return False
+
+        return EditProductionObjects.user_has(
```
Comment (JElgar) [inline uobtheatre/productions/schema.py:395]: Consider existing production on update

File: TODO.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
+Me
+- Fix pylint stuff on alex's branch
```
Comment (atoff) [inline TODO.md]: Remove this file

File: uobtheatre/images/schema.py [hunk 5]
```diff
@@ -16,3 +16,10 @@ class Meta:
         model = Image
         interfaces = (relay.Node,)
         fields = ("id", "url", "alt_text")
+
+
+class Query(graphene.ObjectType):
+    images = graphene.List(ImageNode)
```
Comment (atoff) [inline uobtheatre/images/schema.py:22]: We don't need this

File: uobtheatre/images/views.py [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    # "You will typically want to use both FormParser and MultiPartParser
+    # together in order to fully support HTML form data."
+    parser_classes = (MultiPartParser, FormParser)
+
+    def post(self, request, *_, **__):
+        file_serializer = ImageSerializer(data=request.data)
+        if not file_serializer.is_valid():
+            return Response(file_serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+        file_serializer.save()
```
Comment (atoff) [inline uobtheatre/images/views.py:21]: Can we rename the image file name to something random here?

**Review by atoff — COMMENTED** [review]
Here you go @JElgar

File: uobtheatre/productions/migrations/0013_alter_production_facebook_event.py [hunk 7]
```diff
@@ -0,0 +1,19 @@
+# Generated by Django 3.2.9 on 2021-12-11 13:49
```
Comment (JElgar) [inline uobtheatre/productions/migrations/0013_alter_production_facebook_event.py:1]: I'm going to add this migration into the refund branch so can we shove this in there. We can probably also do some squashing

Comment (JElgar) [reply uobtheatre/productions/migrations/0013_alter_production_facebook_event.py:1]: I also can't find the migration for post_Image but it seems thats changed as welll. I think the best plan is to: 
- delete all migrations in this branch
- merge with dev 
- regenerate migrations

**Review by JElgar — CHANGES_REQUESTED** [review]

File: uobtheatre/discounts/schema.py [hunk 8]
```diff
@@ -18,15 +34,157 @@ class Meta:
...
+    """Delete a discount"""
+
+    @classmethod
+    def authorize_request(cls, _, info, **inputs):
+        discount = cls.get_instance(inputs["id"])
+        for performance in discount.performances.prefetch_related("production").all():
+            if not EditProduction.user_has_for(
+                info.context.user, performance.production
+            ):
+                raise AuthorizationException
```
Comment (JElgar) [inline uobtheatre/discounts/schema.py:121]:
```suggestion
                raise AuthorizationException("You do not have the ability to edit one of the provided performances")
```

File: uobtheatre/discounts/schema.py [hunk 9]
```diff
@@ -18,15 +34,157 @@ class Meta:
...
+        # Authorize user on all of the performance's productions
+        for performance in all_performances:
+            if not EditProduction.user_has_for(
+                info.context.user, performance.production
+            ):
+                raise AuthorizationException(
+                    "You do not have permission to edit one or more of the productions associated with this requirement"
+                )
+
+        return True
```
Comment (JElgar) [inline uobtheatre/discounts/schema.py:156]:
```suggestion
        return
```

File: uobtheatre/discounts/schema.py [hunk 10]
```diff
@@ -18,15 +34,157 @@ class Meta:
...
+    @classmethod
+    def authorize_request(cls, _, info, **inputs):
+        instance = cls.get_instance(inputs["id"])
+        for performance in instance.discount.performances.prefetch_related(
+            "production"
+        ).all():
+            if not EditProduction.user_has_for(
+                info.context.user, performance.production
+            ):
+                raise AuthorizationException()
```
Comment (JElgar) [inline uobtheatre/discounts/schema.py:174]:
```suggestion
                raise AuthorizationException("You do not have the ability to edit one of the provided performances")
```

File: uobtheatre/productions/abilities.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
+from uobtheatre.users.abilities import Ability
+
+
+class AddProduction(Ability):
+    """Whether the user is able to change a production's details and sub-objects"""
```
Comment (JElgar) [inline uobtheatre/productions/abilities.py:5]: Not quite

File: uobtheatre/productions/mutations.py [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+            return has_perm_new_performance and new_performance
+
+        # For update operations, we care that the user has permissions on both the currently assigned performance, and the new performance (if provided)
+        current_performance = cls.get_object_instance(
+            root, info, **inputs
+        ).performance.production
+        has_perm_current_performance = EditProduction.user_has_for(
+            info.context.user, current_performance
+        )
+        return has_perm_new_performance and has_perm_current_performance
```
Comment (JElgar) [inline uobtheatre/productions/mutations.py:271]: Same again

File: uobtheatre/productions/schema.py [hunk 13]
```diff
@@ -19,11 +19,15 @@
...
 from uobtheatre.utils.schema import (
+    AssignedUsersMixin,
     DjangoObjectType,
     GrapheneEnumMixin,
+    IdInputField,
     UserPermissionFilterMixin,
 )
 
+ProductionStatusSchema = graphene.Enum.from_enum(Production.Status)
+
```
Comment (JElgar) [inline uobtheatre/productions/schema.py:30]: Do we need this?

File: uobtheatre/productions/schema.py [hunk 14]
```diff
@@ -355,12 +364,18 @@ class Query(graphene.ObjectType):
...
+    )
     performance = relay.Node.Field(PerformanceNode)
 
-    def resolve_production(self, info, slug):
+    def resolve_production(
+        self, info, id=None, slug=None
+    ):  # pylint: disable=redefined-builtin
         try:
-            return Production.objects.user_can_see(info.context.user).get(slug=slug)
+            qs = Production.objects.user_can_see(info.context.user)
```
Comment (JElgar) [inline uobtheatre/productions/schema.py:378]: Im cleaning this up a bit

File: uobtheatre/productions/test/test_flow.py [hunk 15]
```diff
@@ -0,0 +1,294 @@
+# pylint: disable=R0914,too-many-statements
```
Comment (JElgar) [inline uobtheatre/productions/test/test_flow.py:1]: Can we rename this to test_system

File: uobtheatre/utils/models.py [hunk 16]
```diff
@@ -25,19 +27,42 @@ class Meta:
...
+            PermissionNode(
+                name=permission.codename,
+                description=permission.name,
+                user_can_assign=any(
+                    user.has_perm(permission, self)
+                    for permission in assignable_permissions[permission.codename]
+                )
+                if permission.codename in assignable_permissions
+                else False,
+            )
```
Comment (JElgar) [inline uobtheatre/utils/models.py:63]: What is going on here?

File: uobtheatre/utils/permissions.py [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    codename = split_perm[1] if len(split_perm) > 1 else split_perm[0]
+
+    perm = Permission.objects.filter(
+        content_type=ContentType.objects.get_for_model(Production),
+    ).get(codename=codename)
+
+    return get_users_with_perms(
+        obj,
+        with_superusers=True,
+        only_with_perms_in="productions.approve_production",
```
Comment (JElgar) [inline uobtheatre/utils/permissions.py:24]: Umm?

File: uobtheatre/utils/permissions.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def get_users_with_perm(permission: str, obj: Model):
+    """Get users with a specific global or object-lebel permission"""
+    # First, get users with that permission out right
+    split_perm = permission.split(".")
+    codename = split_perm[1] if len(split_perm) > 1 else split_perm[0]
+
+    perm = Permission.objects.filter(
+        content_type=ContentType.objects.get_for_model(Production),
```
Comment (JElgar) [inline uobtheatre/utils/permissions.py:18]: ?

File: uobtheatre/utils/schema.py [hunk 19]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+            for (user, permissions) in get_users_with_perms(
+                self, attach_perms=True, with_group_users=False
+            ).items()
+        ]
+
+    def resolve_assignable_permissions(self, info):
+        if not info.context.user.has_perm("change_" + str(self._meta.model_name), self):
+            return None
+
+        if not isinstance(self, PermissionableModel):  # pragma: no cover
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:99]: Can we test this?

File: uobtheatre/utils/schema.py [hunk 20]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+
+
+class SafeFormMutation(SafeMutation, DjangoModelFormMutation):
+    """Provides a wrapper around the DjangoModelFormMutation for use with out validation and authorization system"""
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    # pylint: disable=C0116,arguments-differ
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:112]:
```suggestion
    # pylint: disable=arguments-differ
```

Lets add `__init_subclass_with_meta__` to the "no-docstring-rgx" in pylintrc

File: uobtheatre/utils/schema.py [hunk 21]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+
+    @classmethod
+    def is_creation(cls, **inputs):
+        return "id" not in inputs
+
+    @classmethod
+    def mutate(cls, root, info, **inputs):
+        input_items = inputs["input"]
+
+        if "id" in input_items:
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:128]: Can we move this comment and make it clear everything below is straight from graphene

File: uobtheatre/utils/schema.py [hunk 22]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+                except TypeError:
+                    pass
+        return super().mutate(root, info, **input_items)
+
+    @classmethod
+    def get_form_kwargs(cls, root, info, **inputs):
+        return super().get_form_kwargs(root, info, **inputs)
+
+    @classmethod
+    # pylint: disable=W0212
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:152]:
```suggestion
    # pylint: disable=protected-access
```

File: uobtheatre/utils/schema.py [hunk 23]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+        if cls.is_creation(**inputs):
+            user_can_add_instance = info.context.user.has_perm(
+                "%s.add_%s" % (app_label, model_name)
+            )
+            if not (
+                user_can_add_instance
+                if not create_ability
+                else create_ability.user_has(info.context.user)
+            ):
+                raise AuthorizationException
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:169]: Can we add a message here?

File: uobtheatre/utils/schema.py [hunk 24]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+        instance = cls.get_object_instance(root, info, **inputs)
+        user_can_change_instance = info.context.user.has_perm(
+            "change_%s" % model_name, instance
+        )
+        if not (
+            user_can_change_instance
+            if not update_ability
+            else update_ability.user_has_for(info.context.user, instance)
+        ):
+            raise AuthorizationException
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:181]: and here

File: uobtheatre/utils/schema.py [hunk 25]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+    def resolve_mutation(cls, root, info, **inputs):
+        # pylint: disable=bad-super-call
+        return super(DjangoModelFormMutation, cls).mutate(root, info, inputs)
+
+    @classmethod
+    def mutate_and_get_payload(cls, root, info, **inputs):
+        """Mutate and get payload override"""
+        response = super().mutate_and_get_payload(root, info, **inputs)
+
+        if len(response.errors):
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:233]:
```suggestion
        if response.errors:
```

File: uobtheatre/utils/schema.py [hunk 26]
```diff
@@ -35,6 +61,192 @@ def mutate(cls, root, info, **inputs):
...
+            exceptions = GQLExceptions(
+                exceptions=[
+                    GQLException(message, field=error.field)
+                    for error in list(response.errors)
+                    for message in error.messages
+                ]
+            )
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:240]: Can we move this to a custom exception? I've done a similar thing for SquareException which takes a square exception and then does the resolve in the resolve method. 
Something like FormException would be nice

File: uobtheatre/utils/schema.py [hunk 27]
```diff
@@ -69,3 +281,201 @@ def parse_value(ast):
...
+        return cls._meta.model.objects.get(id=model_id)
+
+    @classmethod
+    # pylint: disable=W0212
+    def authorize_request(cls, _, info, **inputs):
+        """Authorize the request"""
+        instance = cls.get_instance(inputs["id"])
+        if cls._meta.ability:
+            if not cls._meta.ability.user_has(info.context.user, instance):
+                raise AuthorizationException
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:326]: Can we add a message

File: uobtheatre/utils/schema.py [hunk 28]
```diff
@@ -69,3 +281,201 @@ def parse_value(ast):
...
+            return
+
+        if not info.context.user.has_perm(
+            "delete_%s" % cls._meta.model._meta.model_name, instance
+        ):
+            raise AuthorizationException
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:332]: same here

File: uobtheatre/utils/schema.py [hunk 29]
```diff
@@ -69,3 +281,201 @@ def parse_value(ast):
...
+        available_permissions = instance.available_permissions_for_user(
+            info.context.user
+        )
+
+        if not info.context.user.has_perm(
+            "change_"
+            + cls._meta.model._meta.model_name,  # pylint: disable=protected-access
+            instance,
+        ):
+            raise AuthorizationException()
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:423]: message

File: uobtheatre/utils/schema.py [hunk 30]
```diff
@@ -69,3 +281,201 @@ def parse_value(ast):
...
+                None,
+            ):
+                raise GQLException(
+                    message="The permission '%s' does not exist" % permission,
+                    field="permissions",
+                )
+
+        if not (user := cls.subject_user(inputs["user_email"])):
+            raise GQLException(
+                "A user with that email does not exist on our system",
```
Comment (JElgar) [inline uobtheatre/utils/schema.py:437]:
```suggestion
                "A user with that email does not exist",
```

**Review by JElgar — APPROVED** [review]
LGMT

**Pull request merged**